### ISSUE TITLE
Skip resolver pipeline function or exit resolver pipeline early

### ISSUE THREAD

**Issue by Angus-Walsh** [aws/aws-appsync-community#261]
Greetings!

I Have been experimenting with the Javascript resolver functions and pipelines, and I was wondering if there was a way to:
- Conditionally passthrough a function in a pipeline, i.e. check if I need to make a request, if not, then pass the previous response to the next function.
- Conditionally exit the pipeline early with a response

As an example, the first resolver function in the pipeline queries a DynamoDB.
I then have the following resolver function that creates a new entry if nothing was found by the previous function.

``` Javascript
function request(ctx) {
  if (ctx.prev.result) {
    return; // Exit the function or pipeline here
  }

  const id = util.autoId();

  return {
    version: '2018-05-29',
    operation: 'PutItem',
    key: util.dynamodb.toMapValues({ id }),
    attributeValues: util.dynamodb.toMapValues(ctx.args)
  };
}

function response(ctx) {
  const { error, result } = ctx;
  if (error) {
    return util.appendError(error.message, error.type, result);
  }
  
  // Check if there was a result, otherwise pass on the previous result (not sure if this would be needed)
  if (result) {
    return result;
  } else {
    return ctx.prev.result;
  }
} 
```

Is it possible in this sort of a situation to "pass through" this resolver, or even return early from the pipeline?

I have seen that in VTL there is the concept of `#return` which looks like what I need but I can't find an equivalent for JS (I could also be misunderstanding `#return` as I am not that familiar with VTL).

**Comment by nikolajwestergaard** [comment]
We have been using #return in VLC, and I have been looking for this functionality. 
Documentation mentions nothing about it!

As we are porting our VLC to APPSYNC_JS I really hope that it is a possibility, in certain circumstances I have to use "cheap calls" to the DB just to get by a unneeded pipeline function 👎

Hope this issue will be prioritised, JavaScript resolvers is really a great leap forward ❤️

**Comment by onlybakam** [comment]
posted a proposal here https://github.com/aws/aws-appsync-community/issues/147
please comment.

**Comment by Hideman85** [comment]
:+1:  This is really needed

**Comment by fanick1** [comment]
I'm interested in this improvement too.  Just exit early when null is returned from the request(), please.

**Comment by nits1991** [comment]
This is really needed. like nikolajwestergaard just returning some cheap operation.

**Comment by onlybakam** [comment]
We now support early return to skip data source request in a function or resolver. See details: https://docs.aws.amazon.com/appsync/latest/devguide/runtime-utils-js.html

**Comment by lqtruong** [comment]
Hi @onlybakam 
Does that e.g. `runtime.earlyReturn` also work with VTL pipeline? Many thanks.

**Comment by micchickenburger** [comment]
> We now support early return to skip data source request in a function or resolver. See details: https://docs.aws.amazon.com/appsync/latest/devguide/runtime-utils-js.html

This seems to only skip the current function execution.  It will NOT stop a pipeline from executing the next function.  Are there any plans for introducing a method for returning early from the entire pipeline?

**Comment by pdramirez-dev** [comment]
Hello @onlybakam  Is there any functionality similar to runtime.earlyReturn but for resolvers in VTL? ?

**Comment by onlybakam** [comment]
Yes. in VTL, you can use [#return](https://docs.aws.amazon.com/appsync/latest/devguide/resolver-mapping-template-reference-overview.html#return-data-object)

**Comment by danielmarzan** [comment]
> > We now support early return to skip data source request in a function or resolver. See details: https://docs.aws.amazon.com/appsync/latest/devguide/runtime-utils-js.html
> 
> This seems to only skip the current function execution. It will NOT stop a pipeline from executing the next function. Are there any plans for introducing a method for returning early from the entire pipeline?

I just tested this and it works, thank you!

**Comment by naedx** [comment]
TO WHOM IT MAY CONCERN:

The signature for `runtime.earlyReturn()` now includes `skipTo` that you can set to 'END' | 'NEXT'.

```
Runtime utils list

runtime.earlyReturn(obj?: unknown, returnOptions?: {skipTo: 'END' | 'NEXT'}): never
Invoking this function will halt the execution of the current handler, AWS AppSync function or resolver (Unit or Pipeline Resolver) depending on the current context. The specified object is returned as the result.

When called in an AWS AppSync function request handler, the data source and response handler are skipped, and the next function request handler (or the pipeline resolver response handler if this was the last AWS AppSync function) is called.

When called in an AWS AppSync pipeline resolver request handler, the pipeline execution is skipped, and the pipeline resolver response handler is called immediately.

When returnOptions is given with skipTo set to "END", the pipeline execution is skipped, and the pipeline resolver response handler is called immediately.

When returnOptions is given with skipTo set to "NEXT", the function execution is skipped, and the next pipeline handler is called.


```

https://docs.aws.amazon.com/appsync/latest/devguide/runtime-utils-js.html

**Comment by naedx** [comment]
The type in the '@aws-appsync/utils' has not been updated to reflect what the documentation describes so you will get a type error locally and in the AppSync console when using options. You cast your function as below as workaround. In the console, the editor will show an error but "Save" will still validate and accept the options.


```typescript
export type EarlyReturnFunctionType = (
  obj?: unknown,
  options?: { skipTo: 'END' | 'NEXT' },
) => never;

return (runtime.earlyReturn as EarlyReturnFunctionType)(
    myValue,
    { skipTo: 'END' },
  );

```

Hi @onlybakam, would you be able to assist with updating the types?
